// artistejs/src/service.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "node_modules/observable/src/index", "node_modules/dependency-injection/src/index", "./mixin"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    const index_1 = require("node_modules/observable/src/index");
    const index_2 = require("node_modules/dependency-injection/src/index");
    const mixin_1 = require("./mixin");
    var injector = new index_2.DependencyInjector();
    exports.config = injector.getConfig();
    exports.serviceProvider = injector.getProvider();
    exports.Service = injector.getDecorator();
    class IServiceProvider {
    }
    exports.IServiceProvider = IServiceProvider;
    class IObservablizer {
    }
    exports.IObservablizer = IObservablizer;
    class INotifier {
    }
    exports.INotifier = INotifier;
    let ServiceProvider = class ServiceProvider extends IServiceProvider {
        getService(type) {
            return exports.serviceProvider.getService(type);
        }
    };
    ServiceProvider = __decorate([
        exports.Service({
            key: IServiceProvider
        })
    ], ServiceProvider);
    let Observablizer = class Observablizer extends IObservablizer {
        convert(value) {
            var res = value && Object.create(value) || undefined;
            value && mixin_1.foreach(value, (item, key) => {
                var descriptor = Object.getOwnPropertyDescriptor(value, key);
                var observable;
                !descriptor.get && !descriptor.set &&
                    (() => {
                        observable = index_1.observable({});
                        descriptor.get = () => observable().value;
                        descriptor.set = (v) => {
                            v instanceof Array && (v.push = function () {
                                var res = Array.prototype.push.apply(this, arguments);
                                observable({ value: this });
                                return res;
                            });
                            v instanceof Array && (v.splice = function () {
                                var res = Array.prototype.splice.apply(this, arguments);
                                observable({ value: this });
                                return res;
                            });
                            observable({ value: v });
                        };
                        delete descriptor.value;
                        delete descriptor.writable;
                        Object.defineProperty(res, key, descriptor);
                        res[key] = item;
                    })();
            });
            return res;
        }
    };
    Observablizer = __decorate([
        exports.Service({
            key: IObservablizer
        })
    ], Observablizer);
    class Event {
        constructor(key) {
            this.key = key;
        }
    }
    exports.Event = Event;
    ;
    let Notifier = class Notifier extends INotifier {
        constructor() {
            super(...arguments);
            this._callbacks = {};
        }
        notify(obj, key, data) {
            var callbacks = this.register(obj, key);
            callbacks && callbacks.forEach((callback) => {
                callback(data);
            });
        }
        listen(obj, key, callback) {
            var callbacks = this.register(obj, key);
            callbacks.push(callback);
        }
        forEvent(event) {
            return {
                listen: (obj, callback) => this.listen(obj, event.key, callback),
                notify: (obj, data) => this.notify(obj, event.key, data)
            };
        }
        register(obj, key) {
            obj.__notifier__id__ = obj.__notifier__id__ || [new Date().getTime(), Math.random() * 100].join("");
            return this._callbacks[obj.__notifier__id__ + "_" + key] = this._callbacks[obj.__notifier__id__ + "_" + key] || [];
        }
    };
    Notifier = __decorate([
        exports.Service({
            key: INotifier
        })
    ], Notifier);
});
